File: python/0045_jump_game_II.py
# n = len(nums)
# Time: O(n)
# Space: O(1)
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        jump_count = 0
        cur_reach = 0
        next_reach = 0

        for idx in range(len(nums) - 1):
            next_reach = max(next_reach, idx + nums[idx])
            if idx == cur_reach:
                jump_count += 1
                cur_reach = next_reach

        return jump_count

File: go/0045_jump_game_II.go
package main

// n = len(nums)
// Time: O(n)
// Space: O(1)
func jump(nums []int) int {
	jumpCount := 0
	curReach := 0
	nextReach := 0

	for i := 0; i < len(nums)-1; i++ {
		nextReach = max(i+nums[i], nextReach)

		if i == curReach {
			jumpCount++
			curReach = nextReach
		}
	}

	return jumpCount
}
